File: microservices/web/src/api/system/sms.ts
import request from '@/utils/request'
import type { PageRequest, PageResponse } from '@/types'

// ---------- 类型 ----------

export type SmsProvider = 'debug' | 'aliyun' | 'tencent'

// 渠道 config：access_key / secret / sign_name 等；密钥服务端脱敏为 ******，
// 编辑时回传 ****** 或留空表示不修改。
export type SmsChannelConfig = Record<string, string>

export interface SmsChannel {
  id: number
  tenant_id: number
  name: string
  provider: SmsProvider
  config?: SmsChannelConfig | null
  status: number
  remark: string
  created_at: string
  updated_at: string
}

export interface SmsTemplate {
  id: number
  tenant_id: number
  code: string
  name: string
  channel_id: number
  content: string
  type: number // 1 验证码 / 2 通知 / 3 营销
  provider_template_id: string
  status: number
  remark: string
  created_at: string
  updated_at: string
}

export type SmsLogStatus = 'sending' | 'success' | 'failure'

export interface SmsLog {
  id: number
  tenant_id: number
  mobile: string
  template_code: string
  content: string
  params?: Record<string, string> | null
  channel_id: number
  channel_name: string
  provider: string
  status: SmsLogStatus
  provider_msg_id: string
  error: string
  created_at: string
  updated_at: string
}

export interface SendSmsResult {
  log_id: number
  status: SmsLogStatus
  content: string
  provider_msg_id?: string
  error?: string
}

export type SmsChannelListParams = PageRequest & {
  keyword?: string
  provider?: string
  status?: number
}

export type SmsTemplateListParams = PageRequest & {
  keyword?: string
  channel_id?: number
  type?: number
  status?: number
}

export type SmsLogListParams = PageRequest & {
  mobile?: string
  template_code?: string
  status?: string
}

export type SmsChannelSaveData = {
  name: string
  provider: SmsProvider
  config?: SmsChannelConfig
  status?: number
  remark?: string
}

export type SmsTemplateSaveData = {
  code: string
  name: string
  channel_id: number
  content: string
  type?: number
  provider_template_id?: string
  status?: number
  remark?: string
}

// ---------- 渠道 ----------

export const getSmsChannelList = (params: SmsChannelListParams) =>
  request.get<unknown, PageResponse<SmsChannel>>('/api/v1/sms/channels', { params })

// 启用中的渠道（模板表单下拉用）
export const getEnabledSmsChannels = () =>
  request.get<unknown, SmsChannel[]>('/api/v1/sms/channels/enabled')

export const createSmsChannel = (data: SmsChannelSaveData) =>
  request.post<unknown, SmsChannel>('/api/v1/sms/channels', data)

export const updateSmsChannel = (id: number, data: Partial<SmsChannelSaveData>) =>
  request.put<unknown, SmsChannel>(`/api/v1/sms/channels/${id}`, data)

export const updateSmsChannelStatus = (id: number, status: number) =>
  request.put<unknown, void>(`/api/v1/sms/channels/${id}/status`, { status })

export const deleteSmsChannel = (id: number) =>
  request.delete<unknown, void>(`/api/v1/sms/channels/${id}`)

// ---------- 模板 ----------

export const getSmsTemplateList = (params: SmsTemplateListParams) =>
  request.get<unknown, PageResponse<SmsTemplate>>('/api/v1/sms/templates', { params })

export const createSmsTemplate = (data: SmsTemplateSaveData) =>
  request.post<unknown, SmsTemplate>('/api/v1/sms/templates', data)

export const updateSmsTemplate = (id: number, data: Partial<SmsTemplateSaveData>) =>
  request.put<unknown, SmsTemplate>(`/api/v1/sms/templates/${id}`, data)

export const updateSmsTemplateStatus = (id: number, status: number) =>
  request.put<unknown, void>(`/api/v1/sms/templates/${id}/status`, { status })

export const deleteSmsTemplate = (id: number) =>
  request.delete<unknown, void>(`/api/v1/sms/templates/${id}`)

// ---------- 发送日志 ----------

export const getSmsLogList = (params: SmsLogListParams) =>
  request.get<unknown, PageResponse<SmsLog>>('/api/v1/sms/logs', { params })

// ---------- 发送（业务发送与模板测试发送共用） ----------

export const sendSms = (data: { mobile: string; template_code: string; params?: Record<string, string> }) =>
  request.post<unknown, SendSmsResult>('/api/v1/sms/send', data)

File: microservices/services/system/internal/model/sms.go
package model

import "time"

// 短信发送状态（sms_logs.status 用字符串，语义比数字直观，方便日志检索）。
const (
	SmsStatusSending = "sending" // 已提交、结果未知
	SmsStatusSuccess = "success" // 云厂商受理成功
	SmsStatusFailure = "failure" // 发送失败（渲染/网络/云厂商拒绝）
)

// 短信模板类型：与前端下拉一致（1 验证码 / 2 通知 / 3 营销）。
const (
	SmsTemplateTypeCode   int8 = 1
	SmsTemplateTypeNotify int8 = 2
	SmsTemplateTypeMarket int8 = 3
)

// SmsChannel 短信渠道：一个渠道对应一组云厂商凭证（租户隔离）。
// Config 里存 access_key / access_secret / sign_name 等；密钥只落库不进代码，
// 对外返回时由 handler 层做脱敏。
type SmsChannel struct {
	ID        uint           `gorm:"primaryKey" json:"id"`
	TenantID  uint           `gorm:"not null;default:1;index" json:"tenant_id"`
	Name      string         `gorm:"size:100;not null" json:"name"`
	Provider  string         `gorm:"size:32;not null" json:"provider"` // debug|aliyun|tencent
	Config    map[string]any `gorm:"column:config;type:json;serializer:json" json:"config"`
	Status    int8           `gorm:"default:1" json:"status"` // 1 启用 / 0 停用
	Remark    string         `gorm:"size:255" json:"remark"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
}

func (SmsChannel) TableName() string {
	return "sms_channels"
}

// SmsTemplate 短信模板：code 租户内唯一；content 含 {name} 形式占位；
// ProviderTemplateID 对应云厂商侧的模板号（阿里 TemplateCode / 腾讯 TemplateID）。
type SmsTemplate struct {
	ID                 uint      `gorm:"primaryKey" json:"id"`
	TenantID           uint      `gorm:"not null;default:1;index" json:"tenant_id"`
	Code               string    `gorm:"size:100;not null" json:"code"`
	Name               string    `gorm:"size:100;not null" json:"name"`
	ChannelID          uint      `gorm:"index" json:"channel_id"`
	Content            string    `gorm:"type:text;not null" json:"content"`
	Type               int8      `gorm:"default:1" json:"type"`
	ProviderTemplateID string    `gorm:"size:100" json:"provider_template_id"`
	Status             int8      `gorm:"default:1" json:"status"`
	Remark             string    `gorm:"size:255" json:"remark"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

func (SmsTemplate) TableName() string {
	return "sms_templates"
}

// SmsLog 发送日志：记录一次发送的手机号、渲染后内容、入参、渠道快照与结果。
type SmsLog struct {
	ID            uint              `gorm:"primaryKey" json:"id"`
	TenantID      uint              `gorm:"not null;default:1;index" json:"tenant_id"`
	Mobile        string            `gorm:"size:32;index" json:"mobile"`
	TemplateCode  string            `gorm:"size:100;index" json:"template_code"`
	Content       string            `gorm:"type:text" json:"content"`
	Params        map[string]string `gorm:"column:params;type:json;serializer:json" json:"params"`
	ChannelID     uint              `json:"channel_id"`
	ChannelName   string            `gorm:"size:100" json:"channel_name"`
	Provider      string            `gorm:"size:32" json:"provider"`
	Status        string            `gorm:"size:16;index" json:"status"` // sending|success|failure
	ProviderMsgID string            `gorm:"size:128" json:"provider_msg_id"`
	Error         string            `gorm:"size:512" json:"error"`
	CreatedAt     time.Time         `json:"created_at"`
	UpdatedAt     time.Time         `json:"updated_at"`
}

func (SmsLog) TableName() string {
	return "sms_logs"
}
